src/data/set/equitable.lean

```diff
@@ -0,0 +1,91 @@
+/-
+Copyright (c) 2021 Yaël Dillies, Bhavik Mehta. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies, Bhavik Mehta
+-/
+import algebra.big_operators.order
+import data.nat.basic
+
+/-!
+# Equitable functions
+
+This file defines equitable functions.
+
+A function `f` is equitable on a set `s` if `f a₁ ≤ f a₂ + 1` for all `a₁, a₂ ∈ s`. This is mostly
+useful when the domain of `f` is `ℕ` or `ℤ` (or more generally a successor order).
+
+## TODO
+
+`ℕ` can be replaced by any `succ_order`, but we don't have those yet.
+-/
+
+open_locale big_operators
+
+variables {α β : Type*}
+
+namespace set
+
+/-- A set is equitable if no element value is more than one bigger than another. -/
+def equitable_on [has_le β] [has_add β] [has_one β] (s : set α) (f : α → β) : Prop :=
+  ∀ ⦃a₁ a₂⦄, a₁ ∈ s → a₂ ∈ s → f a₁ ≤ f a₂ + 1
+
+@[simp]
+lemma equitable_on_empty [has_le β] [has_add β] [has_one β] (f : α → β) :
+  equitable_on ∅ f :=
+λ a _ ha, (set.not_mem_empty _ ha).elim
+
+lemma equitable_on_iff_exists_le_le_add_one {s : set α} {f : α → ℕ} :
+  s.equitable_on f ↔ ∃ b, ∀ a ∈ s, b ≤ f a ∧ f a ≤ b + 1 :=
+begin
+  refine ⟨_, λ ⟨b, hb⟩ x y hx hy, (hb x hx).2.trans (add_le_add_right (hb y hy).1 _)⟩,
+  obtain rfl | ⟨x, hx⟩ := s.eq_empty_or_nonempty,
+  { simp },
+  intros hs,
+  by_cases h : ∀ y ∈ s, f x ≤ f y,
+  { exact ⟨f x, λ y hy, ⟨h _ hy, hs hy hx⟩⟩ },
+  push_neg at h,
+  obtain ⟨w, hw, hwx⟩ := h,
+  refine ⟨f w, λ y hy, ⟨nat.le_of_succ_le_succ _, hs hy hw⟩⟩,
+  rw (nat.succ_le_of_lt hwx).antisymm (hs hx hw),
+  exact hs hx hy,
+end
+
+lemma equitable_on_iff_exists_eq_eq_add_one {s : set α} {f : α → ℕ} :
+  s.equitable_on f ↔ ∃ b, ∀ a ∈ s, f a = b ∨ f a = b + 1 :=
+by simp_rw [equitable_on_iff_exists_le_le_add_one, nat.le_le_add_one_iff]
+
+end set
+
+open set
+
+namespace finset
+
+lemma equitable_on_iff_le_le_add_one {s : finset α} {f : α → ℕ} :
+  equitable_on (s : set α) f ↔
+    ∀ a ∈ s, (∑ i in s, f i) / s.card ≤ f a ∧ f a ≤ (∑ i in s, f i) / s.card + 1 :=
```

Comment:
It could be nice to have fintype/set.finite versions of these too. Also, perhaps it might be helpful to have an RHS of the form `f '' s \subs Icc _ _` here and `f '' s \subs {_, _}` for the below? Otherwise looks good to me, but another maintainer should review this too.

Reply:
What do you mean by the `fintype`/`set.finite` variations? Is it `∑ i in hs.to_finset,`? and `∑ i,`?

Reply:
Yeah, exactly